node/overseer/src/lib.rs

```diff
@@ -964,25 +945,29 @@ where
 			self.on_head_activated(&block.hash);
 		}
 
+		self.clean_up_external_listeners();
+
 		self.broadcast_signal(OverseerSignal::ActiveLeaves(update)).await?;
 
 		Ok(())
 	}
 
 	async fn block_finalized(&mut self, block: BlockInfo) -> SubsystemResult<()> {
 		let mut update = ActiveLeavesUpdate::default();
-		let metrics = &self.metrics;
 
 		self.active_leaves.retain(|h, n| {
 			if *n <= block.number {
 				update.deactivated.push(*h);
-				metrics.on_head_deactivated();
 				false
 			} else {
 				true
 			}
 		});
 
+		for deactivated in &update.deactivated {
```

Comment:
I take it this is a separate loop instead of happening within the `active_leaves.retain` call for borrowck reasons?

Reply:
exactly